sync/settings.go

```diff
@@ -0,0 +1,305 @@
+		if err := json.Unmarshal(ar.InputData, request); err != nil {
+			return nil, fmt.Errorf("%w: failed to parse import request: %s", ErrInvalidImport, err)
+		}
+	}
+
+	// Check if we need to parse the export.
+	switch {
+	case request.Export != nil && request.RawExport != "":
+		return nil, fmt.Errorf("%w: both Export and RawExport are defined", ErrInvalidImport)
+	case request.RawExport != "":
+		// TODO: Verify checksum for integrity.
+
+		export := &SettingsExport{}
+		if err := yaml.Unmarshal([]byte(request.RawExport), export); err != nil {
+			return nil, fmt.Errorf("%w: failed to parse export: %s", ErrInvalidImport, err)
+		}
+		request.Export = export
+	}
+
+	// Import.
+	return ImportSettings(request)
+}
+
+// ExportSettings exports the global settings.
+func ExportSettings(from string) (*SettingsExport, error) {
+	var settings map[string]any
+	if from == ExportTargetGlobal {
+		// Collect all changed global settings.
+		settings = make(map[string]any)
+		_ = config.ForEachOption(func(option *config.Option) error {
+			v := option.UserValue()
+			if v != nil {
+				settings[option.Key] = v
+			}
+			return nil
+		})
+	} else {
+		r, err := db.Get(profile.ProfilesDBPath + from)
+		if err != nil {
+			return nil, fmt.Errorf("%w: failed to find profile: %s", ErrTargetNotFound, err)
+		}
+		p, err := profile.EnsureProfile(r)
+		if err != nil {
+			return nil, fmt.Errorf("%w: failed to load profile: %s", ErrExportFailed, err)
+		}
+		settings = config.Flatten(p.Config)
+	}
+
+	// Check if there any changed settings.
+	if len(settings) == 0 {
+		return nil, ErrUnchanged
+	}
+
+	// Expand config to hierarchical form.
+	settings = config.Expand(settings)
+
+	return &SettingsExport{
+		Type:   TypeSettings,
+		Config: settings,
+	}, nil
+}
+
+// ImportSettings imports the global settings.
+func ImportSettings(r *SettingsImportRequest) (*ImportResult, error) {
+	// Check import.
+	if r.Export.Type != TypeSettings {
+		return nil, ErrMismatch
+	}
+
+	// Flatten config.
+	settings := config.Flatten(r.Export.Config)
+
+	// Validate config and gather some metadata.
+	var (
+		result  = &ImportResult{}
+		checked int
+	)
+	err := config.ForEachOption(func(option *config.Option) error {
+		// Check if any setting is set.
+		if r.Reset && option.IsSetByUser() {
+			result.ReplacesExisting = true
+		}
+
+		newValue, ok := settings[option.Key]
+		if ok {
+			checked++
+
+			// Validate the new value.
+			if err := option.ValidateValue(newValue); err != nil {
+				return fmt.Errorf("%w: configuration value for %s is invalid: %s", ErrInvalidSetting, option.Key, err)
+			}
+
+			// Collect metadata.
+			if option.RequiresRestart {
+				result.RestartRequired = true
+			}
+			if !r.Reset && option.IsSetByUser() {
+				result.ReplacesExisting = true
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	if checked < len(settings) {
+		return nil, fmt.Errorf("%w: the export contains unknown settings", ErrInvalidImport)
+	}
```

Comment:
This will break the plugin system when it eventually get's merged

Reply:
I would expect the plugins to be already loaded when settings are imported.

We could however, treat this as a warning instead - but we can also "weaken" this later on.